File: src/main/java/nl/sugcube/dirtyarrows/region/RegionManager.java
package nl.sugcube.dirtyarrows.region;

import nl.sugcube.dirtyarrows.DirtyArrows;
import nl.sugcube.dirtyarrows.util.DaUtil;
import org.bukkit.Location;

import java.util.ArrayList;
import java.util.List;

public class RegionManager {

	public static DirtyArrows plugin;
	
	private final List<Region> registeredRegions = new ArrayList<>();
	private Location position1;
	private Location position2;
	
	public RegionManager(DirtyArrows da) {
		plugin = da;
	}
	
	/**
	 * Removes a region.
	 * @param name (String) Name of the region to remove.
	 */
	public void removeRegion(String name) {
		for (int i = 0; i < registeredRegions.size(); i++) {
			Region reg = registeredRegions.get(i);
			if (reg.getName().equalsIgnoreCase(name)) {
				registeredRegions.remove(reg);
				plugin.getData().set("regions." + name, null);
				plugin.saveData();
				plugin.reloadData();
				return;
			}
		}
	}
	
	/**
	 * Sets the selected location of position 1 or 2.
	 * @param id (int) 1 for position 1, and 2 for position 2
	 * @param loc (Location) Location to set the selection to.
	 */
	public void setSelection(int id, Location loc) {
		if (id == 1) {
			this.position1 = loc;
		} else {
			this.position2 = loc;
		}
	}
	
	/**
	 * Loads all the regions from the data.yml
	 */
	public void loadRegions() {
		if (!plugin.getData().isSet("regions")) {
			return;
		}
		for (String key : plugin.getData().getConfigurationSection("regions").getKeys(false)) {
			String path = "regions." + key + ".";
			String pos1string = plugin.getData().getString(path + "pos1");
			String pos2string = plugin.getData().getString(path + "pos2");
			Location loc1 = DaUtil.toLocation(pos1string);
			Location loc2 = DaUtil.toLocation(pos2string);
			Region region = new Region(loc1, loc2, key);
			registeredRegions.add(region);
		}
	}
	
	/**
	 * Saves all the regions to the data.yml
	 */
	public boolean saveRegions() {
		boolean error = false;
        for (Region registeredRegion : registeredRegions) {
            try {
                if (registeredRegion == null || registeredRegion.getName() == "") {
                    continue;
                }

                if (registeredRegion.getLocation(1) != null) {
                    plugin.getData().set("regions." + registeredRegion.getName() + ".pos1", DaUtil.toLocationString(registeredRegion.getLocation(1)));
                }

                if (registeredRegion.getLocation(2) != null) {
                    plugin.getData().set("regions." + registeredRegion.getName() + ".pos2", DaUtil.toLocationString(registeredRegion.getLocation(2)));
                }

                plugin.saveData();
            }
            catch (Exception e) {
                error = true;
            }
        }
		plugin.getLogger().info("Could not save locations.");
        return error;
	}
	
	/**
	 * Gives a list of all region-names stored.
	 * @return (StringList)
	 */
	public List<String> getAllNames() {
		List<String> names = new ArrayList<String>();
        for (Region registeredRegion : registeredRegions) {
            names.add(registeredRegion.getName());
        }
		
		return names;
	}
	
	/**
	 * Checks if there already exists a region with the given name.
	 * @param name (String) The region's name to be checked.
	 * @return (boolean) True if it does exist, false if not.
	 */
	public boolean doesExist(String name) {
        for (Region registeredRegion : registeredRegions) {
            if (registeredRegion.getName().equalsIgnoreCase(name)) {
                return true;
            }
        }
		
		return false;
	}
	
	/**
	 * Checks if there already exists a region with the given name.
	 * @param reg (Region) The region to be checked.
	 * @return (boolean) True if it does exist, false if not.
	 */
	public boolean doesExist(Region reg) {
        for (Region registeredRegion : registeredRegions) {
            if (registeredRegion.getName().equalsIgnoreCase(reg.getName())) {
                return true;
            }
        }
		
		return false;
	}
	
	/**
	 * Gets a region by it's name.
	 * @param name (String) name of the Region.
	 * @return (Region) or `null` when there is no region with the given name.
	 */
	public Region regionByName(String name) {
        for (Region registeredRegion : registeredRegions) {
            if (registeredRegion.getName().equalsIgnoreCase(name)) {
                return registeredRegion;
            }
        }
		
		return null;
	}
	
	public Region isWithinAXZMargin(Location loc, int margin) {
        for (Region region : registeredRegions) {
            if (RegionManager.isWithinXZMargin(region, loc, margin)) {
                return region;
            }
        }
		
		return null;
	}
	
	/**
	 * Checks wether the given locations is within one of the
	 * registered regions (with margin).
	 * @param loc (Location)
	 * @param margin (int) expansion of the region.
	 * @return (Region) null if not in a region.
	 */
	public Region isWithinARegionMargin(Location loc, int margin) {
        for (Region region : registeredRegions) {
            if (RegionManager.isWithinRegionMargin(region, loc, margin)) {
                return region;
            }
        }
		
		return null;
	}
	
	/**
	 * Checks wether the given location is within one of the
	 * registered regions.
	 * @param loc (Location)
	 * @return (Region) null if not in a region.
	 */
	public Region isWithinARegion(Location loc) {
        for (Region region : registeredRegions) {
            if (RegionManager.isWithinRegion(region, loc)) {
                return region;
            }
        }
		
		return null;
	}
	
	/**
	 * Creates a region from the positions stored in memory and will
	 * immedeately register the region.
	 * @param registerName (String) The name (ID) of the region.
	 * @return (Region) The created region.
	 */
	public Region createRegion(String registerName) {
		if (position1 != null && position2 != null) {
			Region region = new Region(position1, position2, registerName);
			this.registeredRegions.add(region);
			return region;
		}
		
		return null;
	}
	
	/**
	 * Creates a region from the positions stored in memory.
	 * @return (Region) The instance created, or null if one or both of
	 * the positions aren't set.
	 */
	public Region createRegion() {
		if (position1 != null && position2 != null) {
            return new Region(position1, position2);
		}
		
		return null;
	}
	
	/**
	 * Checks wether the given locations is within the region expanded with
	 * a certain margin.
	 * @param region (Region)
	 * @param loc (Location)
	 * @param margin (int) The amount of blocks outside the area to check for.
	 * @return (boolean) True if within region with margin, false if not.
	 */
	public static boolean isWithinXZMargin(Region region, Location loc, int margin) {
		double p1X = region.getLocation(1).getX();
		double p1Z = region.getLocation(1).getZ();
		double p2X = region.getLocation(2).getX();
		double p2Z = region.getLocation(2).getZ();
		
		if (p1X <= p2X) {
			p1X -= margin;
			p2X += margin;
		} else {
			p1X += margin;
			p2X -= margin;
		}
		
		if (p1Z <= p2Z) {
			p1Z -= margin;
			p2Z += margin;
		} else {
			p1Z += margin;
			p2Z -= margin;
		}
		
		if (!loc.getWorld().equals(loc.getWorld())) {
			return false;
		}
		
		if ((p2X <= loc.getX() && loc.getX() < p1X) || (p1X <= loc.getX() && loc.getX() < p2X)) {
			if ((p2Z <= loc.getZ() && loc.getZ() < p1Z) || (p1Z <= loc.getZ() && loc.getZ() < p2Z)) {
				return true;
			}
		}
		
		return false;
	}
	
	/**
	 * Checks wether the given locations is within the region expanded with
	 * a certain margin.
	 * @param region (Region)
	 * @param loc (Location)
	 * @param margin (int) The amount of blocks outside the area to check for.
	 * @return (boolean) True if within region with margin, false if not.
	 */
	public static boolean isWithinRegionMargin(Region region, Location loc, int margin) {
		double p1X = region.getLocation(1).getX();
		double p1Y = region.getLocation(1).getY();
		double p1Z = region.getLocation(1).getZ();
		double p2X = region.getLocation(2).getX();
		double p2Y = region.getLocation(2).getY();
		double p2Z = region.getLocation(2).getZ();
		
		if (p1X <= p2X) {
			p1X -= margin;
			p2X += margin;
		} else {
			p1X += margin;
			p2X -= margin;
		}
		
		if (p1Y <= p2Y) {
			p1Y -= margin;
			p2Y += margin;
		} else {
			p1Y += margin;
			p2Y -= margin;
		}
		
		if (p1Z <= p2Z) {
			p1Z -= margin;
			p2Z += margin;
		} else {
			p1Z += margin;
			p2Z -= margin;
		}
		
		if (!loc.getWorld().equals(loc.getWorld())) {
			return false;
		}
		
		if ((p2X <= loc.getX() && loc.getX() < p1X) || (p1X <= loc.getX() && loc.getX() < p2X)) {
			if ((p2Y <= loc.getY() && loc.getY() < p1Y) || (p1Y <= loc.getY() && loc.getY() < p2Y)) {
				if ((p2Z <= loc.getZ() && loc.getZ() < p1Z) || (p1Z <= loc.getZ() && loc.getZ() < p2Z)) {
					return true;
				}
			}
		}
		
		return false;
	}
	
	/**
	 * Checks wether the given locations is within the specified region.
	 * @param region (Region)
	 * @param loc (Location)
	 * @return (boolean) True if within, false if not.
	 */
	public static boolean isWithinRegion(Region region, Location loc) {
		double p1X = region.getLocation(1).getX();
		double p1Y = region.getLocation(1).getY();
		double p1Z = region.getLocation(1).getZ();
		double p2X = region.getLocation(2).getX();
		double p2Y = region.getLocation(2).getY();
		double p2Z = region.getLocation(2).getZ();
		
		if (!region.getLocation(1).getWorld().equals(loc.getWorld())) {
			return false;
		}
		
		if ((p2X <= loc.getX() && loc.getX() < p1X) || (p1X <= loc.getX() && loc.getX() < p2X)) {
			if ((p2Y <= loc.getY() && loc.getY() < p1Y) || (p1Y <= loc.getY() && loc.getY() < p2Y)) {
				if ((p2Z <= loc.getZ() && loc.getZ() < p1Z) || (p1Z <= loc.getZ() && loc.getZ() < p2Z)) {
					return true;
				}
			}
		}
		
		return false;
	}

	public List<Region> getRegisteredRegions() {
		return registeredRegions;
	}
	
}

File: src/main/kotlin/nl/sugcube/dirtyarrows/region/RegionManager.kt
package nl.sugcube.dirtyarrows.region

import nl.sugcube.dirtyarrows.DirtyArrows
import nl.sugcube.dirtyarrows.util.toLocation
import nl.sugcube.dirtyarrows.util.toLocationString
import org.bukkit.Location
import java.util.logging.Level

/**
 * @author SugarCaney
 */
open class RegionManager(private val plugin: DirtyArrows) {

    /**
     * Contains all known regions. Maps the region name/id to the actual region object.
     */
    private val registeredRegions = HashMap<String, Region>()

    /**
     * The first selected position.
     */
    private var position1: Location? = null

    /**
     * The second selected position.
     */
    private var position2: Location? = null

    /**
     * Get all registered regions.
     */
    val allRegions: List<Region>
        get() = registeredRegions.values.toList()

    /**
     * The names of all registered regions.
     */
    val allNames: Collection<String>
        get() = registeredRegions.keys.toList()

    /**
     * Get the amount of regions that have been registered.
     */
    val size: Int
        get() = registeredRegions.size

    /**
     * Checks whether a region is registered with the given name.
     */
    fun regionExists(name: String): Boolean = regionByName(name) != null

    /**
     * Get the registered region with the given name.
     *
     * @param name
     *          The name of the region to lookup.
     * @return The region or `null` if there is no region with the given name.
     */
    fun regionByName(name: String) = registeredRegions[name]

    /**
     * Creates a region and register it to the Region Manager.
     * Does not update the data configuration.
     *
     * @param registerName
     *          The name of the region to create, must only contain A-Z, a-z, 0-9, _, characters.
     * @return The created region, or `null` when the region could not be created.
     */
    fun createRegion(registerName: String): Region? {
        if (position1 == null || position2 == null) return null
        require(Region.NAME_REGEX.matches(registerName)) { "Invalid region name: '$registerName'" }

        val region = Region(position1!!, position2!!, registerName)
        registeredRegions[registerName] = region
        return region
    }

    /**
     * Unregisters the given region and also removes it from the data file.
     */
    fun removeRegion(name: String) {
        if (regionExists(name).not()) return
        registeredRegions.remove(name)
        plugin.data.set("regions.$name", null)
        plugin.configurationManager.saveData()
    }

    /**
     * Gets position #`positionNumber`.
     *
     * @param positionNumber
     *          Either 1 or 2 for position 1 and 2 respectively.
     * @return The set position, or `null` when the position hasn't been set.
     */
    fun getSelection(positionNumber: Int) = when (positionNumber) {
        1 -> position1
        2 -> position2
        else -> error("Invalid location number '$positionNumber', only 1|2 allowed.")
    }

    /**
     * Sets position #`positionNumber` to the given location.
     *
     * @param positionNumber
     *          Either 1 or 2 for position 1 and 2 respectively.
     * @param location
     *          The location to set the position to.
     */
    fun setSelection(positionNumber: Int, location: Location) = when (positionNumber) {
        1 -> position1 = location
        2 -> position2 = location
        else -> error("Invalid location number '$positionNumber', only 1|2 allowed.")
    }

    /**
     * Loads all the regions from the data configuration and registers them.
     * Overwrites all current regions.
     */
    fun loadRegions() {
        val data = plugin.data

        if (data.isSet("regions").not()) {
            registeredRegions.clear()
            return
        }

        data.getConfigurationSection("regions").getKeys(false).forEach { key ->
            val path = "regions.$key"
            val position1 = data.getString("$path.pos1").toLocation() ?: return@forEach
            val position2 = data.getString("$path.pos2").toLocation() ?: return@forEach
            val region = Region(position1, position2, key)
            registeredRegions[key] = region
        }

        plugin.logger.log(Level.INFO, "Loaded ${registeredRegions.size} regions.")
    }

    /**
     * Saves all registered regions to the data.yml
     */
    fun saveRegions() {
        val data = plugin.data
        registeredRegions.forEach { (regionName, region) ->
            val location1 = region.getLocation(1).toLocationString()
            val location2 = region.getLocation(2).toLocationString()

            data.set("regions.$regionName.pos1", location1)
            data.set("regions.$regionName.pos2", location2)
        }

        plugin.configurationManager.saveData()

        plugin.logger.log(Level.INFO, "Saved ${registeredRegions.size} regions to data.yml.")
    }

    /**
     * Checks whether the given location is within a region + margin.
     * Not counting height.
     *
     * @return The first region the location is in, or `null` if the location is not within a region + margin.
     */
    fun isWithinARegionXZMargin(location: Location, margin: Double): Region? {
        return registeredRegions.values.firstOrNull { it.isWithinXZMargin(location, margin) }
    }

    /**
     * Checks whether the given location is within a region + margin.
     *
     * @return The first region the location is in, or `null` if the location is not within a region + margin.
     */
    fun isWithinARegionMargin(location: Location, margin: Double): Region? {
        return registeredRegions.values.firstOrNull { it.isWithinMargin(location, margin) }
    }

    /**
     * Checks wether the given location is within one of the registered regions.
     *
     * @return The first region the location is in, or `null` if the location is not within a region.
     */
    fun isWithinARegion(location: Location) = isWithinARegionMargin(location, margin = 0.0)
}